x-pack/plugin/identity-provider/src/main/java/org/elasticsearch/xpack/idp/saml/authn/SamlAuthnRequestValidator.java

```diff
@@ -73,76 +74,108 @@ public SamlAuthnRequestValidator(SamlFactory samlFactory, SamlIdentityProvider i
-                            new ParameterizedMessage("Unable to validate signature of authentication request [{}] using credentials [{}]",
-                            queryString, samlFactory.describeCredentials(Collections.singletonList(spSigningCredential))), listener);
-                        return;
-                    }
-                } else if (Strings.hasText(sigAlg)) {
-                    logAndRespond(new ParameterizedMessage("Query string [{}] contains a SigAlg parameter but Signature is missing",
-                        queryString), listener);
-                    return;
-                } else {
-                    logAndRespond("The Service Provider must sign authentication requests but no signature was found", listener);
-                    return;
-                }
-            }
-            validateAuthnRequest(authnRequest, sp);
-            Map<String, Object> authnState = buildAuthnState(authnRequest, sp);
-            final SamlValidateAuthnRequestResponse response = new SamlValidateAuthnRequestResponse(sp.getEntityId(),
-                authnRequest.isForceAuthn(), authnState);
-            logger.trace(new ParameterizedMessage("Validated AuthnResponse from queryString [{}] and extracted [{}]",
-                queryString, response));
-            listener.onResponse(response);
+                },
+                listener::onFailure
+            ));
         } catch (ElasticsearchSecurityException e) {
-            logger.debug("Could not process and validate AuthnRequest", e);
+            logger.debug("Could not process AuthnRequest", e);
             listener.onFailure(e);
         } catch (Exception e) {
-            logAndRespond("Could not process and validate AuthnRequest", e, listener);
+            logAndRespond("Could not process AuthnRequest", e, listener);
+        }
+    }
+
+    private ParsedQueryString parseQueryString(String queryString) throws ElasticsearchSecurityException {
+
+        final Map<String, String> parameters = new HashMap<>();
+        RestUtils.decodeQueryString(queryString, 0, parameters);
+        if (parameters.isEmpty()) {
+            throw new ElasticsearchSecurityException("Invalid Authentication Request query string (zero parameters)");
+        }
+        logger.trace(new ParameterizedMessage("Parsed the following parameters from the query string: {}", parameters));
+        final String samlRequest = parameters.get("SAMLRequest");
+        if (null == samlRequest) {
+            throw new ElasticsearchSecurityException("Query string [{}] does not contain a SAMLRequest parameter", queryString);
         }
+        return new ParsedQueryString(
+            queryString,
+            samlRequest,
+            parameters.get("RelayState"),
+            parameters.get("SigAlg"),
+            parameters.get("Signature"));
+    }
+
+    private boolean validateAuthnRequest(AuthnRequest authnRequest, SamlServiceProvider sp, ParsedQueryString parsedQueryString,
```

Comment:
TODO: This return value (`boolean`) is obsolete. It was there from IntelliJ's refactoring and I forgot to fix it.

Reply:
ack